lib/pbench/server/cache_manager.py

```diff
@@ -1433,3 +1505,124 @@ def delete(self, dataset_id: str):
         del self.tarballs[name]
 
         self._clean_empties(tarball.controller_name)
+
+    def reclaim_cache(self, goal_pct: float = 20.0, goal_bytes: int = 0) -> bool:
+        """Reclaim unused caches until at least goal (%) disk is available.
+
+        The cache tree need not be fully discovered at this point; we can
+        reclaim cache even on a partial tree. This is driven by discovery of
+        the cache directory tree, looking for <resource_id> directories that
+        contain an unpacked tarball root rather than only the `lock` and
+        `last_ref` files.
+
+        This is a "best effort" operation. It will free least recently used
+        caches that are currently unlocked until both the goal % is achieved
+        and the absolute goal bytes is achieved (if specified), or until there
+        are no more unlocked cached tarballs.
+
+        Args:
+            tree: the cache manager instance
+            goal_pct: goal percent of cache filesystem free
+            goal_bytes: goal in bytes
+
+        Returns:
+            True if both goals are met, or False otherwise
+        """
+
+        @dataclass
+        class Candidate:
+            last_ref: float  # last reference timestamp
+            cache: Path  # Unpacked artifact directory
+
+        def reached_goal():
+            usage = shutil.disk_usage(self.cache_root)
+            available = float(usage.free) / float(usage.total) * 100.0
+            return bool(available >= goal_pct and usage.free >= goal_bytes)
+
+        if reached_goal():
+            return True
+
+        total_count = 0
+        has_cache = 0
+        reclaimed = 0
+        reclaim_failed = 0
+        self.logger.info(
+            "RECLAIM: looking for {}% or {} free",
+            goal_pct,
+            humanize.naturalsize(goal_bytes),
+        )
+
+        # Identify cached datasets by examining the cache directory tree
+        candidates = deque()
+        for d in self.cache_root.iterdir():
+            if not d.is_dir():
+                self.logger.warning("unexpected file in cache root: {}", d)
+                continue
+            total_count += 1
+            last_ref = 0.0
+            for f in d.iterdir():
+                if f.name == "last_ref":
+                    last_ref = f.stat().st_mtime
+                elif f.is_dir():
+                    candidates.append(Candidate(last_ref, f))
+                    break
```

Comment:
No. 🐑 And it's not worth the complication of trying to track both so I guess I'll just drop the `break`. There should only be 2 files and at most one directory anyway...

Reply:
Since instantiating `Candidate` requires `last_ref`, it would be good if we had that value _before_ we instantiated it.

So, it might be better to find `f` and find `last_ref` and _then_ instantiate the `Candidate`.